File: src/main/java/com/estore/estore/service/CategoryService.java
package com.estore.estore.service;

import com.estore.estore.dto.request.CategoryRequest;
import com.estore.estore.model.Category;
import com.estore.estore.repository.CategoryRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.Optional;

@Service
public class CategoryService {

    @Autowired
    private CategoryRepository categoryRepository;

    public List<Category> getAllCategories() {
        return categoryRepository.findAllByOrderByNameAsc();
    }

    public Optional<Category> getCategoryById(Long id) {
        return categoryRepository.findById(id);
    }

    public Category createCategory(CategoryRequest categoryRequest) {
        // Проверяем, нет ли уже категории с таким именем
        if (categoryRepository.existsByName(categoryRequest.getName())) {
            throw new RuntimeException("Category with name '" + categoryRequest.getName() + "' already exists");
        }

        // Создаем категорию из DTO
        Category category = new Category();
        category.setName(categoryRequest.getName());
        category.setDescription(categoryRequest.getDescription());

        return categoryRepository.save(category);
    }

    public Category updateCategory(Long id, CategoryRequest categoryRequest) {
        Category category = categoryRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Category not found with id: " + id));

        category.setName(categoryRequest.getName());
        category.setDescription(categoryRequest.getDescription());

        return categoryRepository.save(category);
    }

    public void deleteCategory(Long id) {
        Category category = categoryRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Category not found with id: " + id));
        categoryRepository.delete(category);
    }
}

File: src/main/resources/static/js/services/category.service.js
// static/js/services/category.service.js - НОВЫЙ ФАЙЛ

class CategoryService {
    /**
     * Получить все категории
     */
    static async getAllCategories() {
        try {
            return await ApiService.get('/categories');
        } catch (error) {
            console.error('Error loading categories:', error);
            
            // Возвращаем тестовые данные при ошибке
            return [
                { id: 1, name: 'Смартфоны', description: 'Мобильные телефоны и гаджеты' },
                { id: 2, name: 'Ноутбуки', description: 'Портативные компьютеры' },
                { id: 3, name: 'Телевизоры', description: 'Телевизоры и мониторы' },
                { id: 4, name: 'Аудиотехника', description: 'Наушники и колонки' },
                { id: 5, name: 'Периферия', description: 'Клавиатуры, мыши, аксессуары' }
            ];
        }
    }

    /**
     * Получить категорию по ID
     */
    static async getCategoryById(id) {
        try {
            return await ApiService.get(`/categories/${id}`);
        } catch (error) {
            console.error(`Error loading category ${id}:`, error);
            return null;
        }
    }

    /**
     * Создать категорию (админ)
     */
    static async createCategory(categoryData) {
        try {
            return await ApiService.post('/categories', categoryData, true);
        } catch (error) {
            console.error('Error creating category:', error);
            throw error;
        }
    }

    /**
     * Обновить категорию (админ)
     */
    static async updateCategory(id, categoryData) {
        try {
            return await ApiService.put(`/categories/${id}`, categoryData, true);
        } catch (error) {
            console.error(`Error updating category ${id}:`, error);
            throw error;
        }
    }

    /**
     * Удалить категорию (админ)
     */
    static async deleteCategory(id) {
        try {
            return await ApiService.delete(`/categories/${id}`, true);
        } catch (error) {
            console.error(`Error deleting category ${id}:`, error);
            throw error;
        }
    }
}

window.CategoryService = CategoryService;